pkg/flow/import_test.go

```diff
@@ -0,0 +1,769 @@
+package flow_test
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/grafana/agent/pkg/flow"
+	"github.com/grafana/agent/pkg/flow/internal/testcomponents"
+	"github.com/grafana/agent/pkg/flow/logging"
+	"github.com/grafana/agent/service"
+	"github.com/stretchr/testify/require"
+
+	_ "github.com/grafana/agent/component/module/string"
+)
+
+func TestImportString(t *testing.T) {
+	declare := `
+        declare "test" {
+            argument "input" {}
+
+            testcomponents.passthrough "pt" {
+                input = argument.input.value
+                lag = "1ms"
+            }
+
+            export "testOutput" {
+                value = testcomponents.passthrough.pt.output
+            }
+        }
+    `
+	testCases := []struct {
+		name   string
+		config string
+	}{
+		{
+			name: "Import declare and instantiate at the root",
+			config: `
+                testcomponents.count "inc" {
+                    frequency = "10ms"
+                    max = 10
+                }
+
+                import.string "testImport" {
+                    content = ` + strconv.Quote(declare) + `
+                }
+
+                testImport.test "myModule" {
+                    input = testcomponents.count.inc.count
+                }
+
+                testcomponents.summation "sum" {
+                    input = testImport.test.myModule.testOutput
+                }
+            `,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer verifyNoGoroutineLeaks(t)
+			ctrl := flow.New(testOptions(t))
+			f, err := flow.ParseSource(t.Name(), []byte(tc.config))
+			require.NoError(t, err)
+			require.NotNil(t, f)
+
+			err = ctrl.LoadSource(f, nil)
+			require.NoError(t, err)
+
+			ctx, cancel := context.WithCancel(context.Background())
+			done := make(chan struct{})
+			go func() {
+				ctrl.Run(ctx)
+				close(done)
+			}()
+			defer func() {
+				cancel()
+				<-done
+			}()
+
+			// Check for initial condition
+			require.Eventually(t, func() bool {
+				export := getExport[testcomponents.SummationExports](t, ctrl, "", "testcomponents.summation.sum")
+				return export.LastAdded == 10
+			}, 3*time.Second, 10*time.Millisecond)
+		})
+	}
+}
+
+func TestImport(t *testing.T) {
```

Comment:
I'm interested in seeing a test case for a deeply nested import where the definition of the import changes and is used to propagate something up to the root controller; I can't tell if you already have a test case that does something like this.

Reply:
I had some tests but it was unreadable. I refactored it with txtar and it should now be much clearer. I do have tests where deeply nested modules are being updated. Let me know what you think about the new structure of the tests and there are more cases that you would like me to cover